### PR TITLE
Add support for math.isclose()

### PR DESCRIPTION
**Pull request by njriasan**
Handles half of #5977. Adds the proper math.isclose support/changes requested from https://github.com/numba/numba/pull/6074. This is done because that PR has been open too long and it became too frustrating to merge.

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
Hey Nick, thanks for adding this function. Just a suggestion but don't feel obligated to add it. The tests used by CPython for `math.isclose` are available here:
https://github.com/python/cpython/blob/56310136170ef6653cb050f58dd2d32538997f59/Lib/test/test_math.py#L2122-L2242

Perhaps one can use some tests to ensure the implementation is correct.

**Comment by stuartarchibald** [comment]
Am going to try out the check list here https://github.com/numba/numba/pull/8209 on this PR, hope that's ok @njriasan?

I have checked that:

* [ ] There is no other Pull Request proposing the same change. (If there is
        please highlight this and one of the core developers will pick it up.)
* [ ] The proposed additional Python functionality is not deprecated by
        Python.
* [ ] Public CI is passing.
* [ ] The patch does not touch files in `numba.core`.
* [ ] The patch is implemented using the high-level extension API only
        (https://numba.readthedocs.io/en/stable/extending/high-level.html).
* [ ] The patch updates the documentation to reflect the new functionality.
        (https://github.com/numba/numba/blob/main/docs/source/reference/pysupported.rst)
* [ ] The patch is line wrapped to 80 characters.
* [ ] Changes to Python source in the patch are `flake8` compliant.
* [ ] The implementation of the Python function reflects that present in the
        Python implementation itself (look in the Python repository:
        https://github.com/python/cpython)...
    * [ ] The implementation has references to the region of the Python code
            base that provides the same functionality. The references are URLs
            of the form:
            `https://github.com/python/cpython/tree/<SHA>/path/to/file.py:#<lines>`
    * [ ] The code "looks" reasonably similar to the Python implementation.
    * [ ] The code uses the same algorithm as the Python implementation, if
            it does not it explains why and this justification is reasonable.
    * [ ] The type inference part of the Numba implementation restricts the
            types to those declared in the Python docs (or if not present,
            what is accepted in the Python implementation itself).
    * [ ] The Numba implementation uses the same variable names for
            arguments as in the Python implementation.
    * [ ] The Numba code has commentary present for anything not-obvious by
            inspection.
    * [ ] The Numba code accommodates differences that might occur across
            different Python versions in a clear fashion and uses
            `numba.core.utils.PYVERSION` to separate the difference in
            implementations.
* [ ] The unit tests for the implementation:
    * [ ] Use the same test inputs as the Python unit tests for the same
            functionality. If this is not the case there are comments which
            explain why and this justification is reasonable.
    * [ ] The Python unit test inputs have reference URLs associated with
            them to track their origin. The references are URLs of the form:
            `https://github.com/python/cpython/tree/<SHA>/path/to/file.py:#<lines>`
    * [ ] Use `unittest` style assertions (not pytest style!).
    * [ ] Check that unsupported/invalid input types are handled as expected
            using the `with self.assertRaises()` pattern.
    * [ ] Explicitly check custom error messages are raised appropriately.
    * [ ] Hit all the branches of the proprosed change (both at type
            inference and run time).
    * [ ] Check that numerically awkward values work as expected e.g.
            `NaN`, `+/-Inf`, `0`, `INT_MAX`, `INT_MIN` etc.
    * [ ] Check that passing empty versions of containers e.g. empty tuple
            or 0d array works as expected.
* [ ] The patch works in a simple manual test in my local checkout.
* [ ] I cannot break the code in the patch in my local checkout after
        trying to do so for a minimum of 10 minutes. The error messages seem
        reasonable and there are no "incorrect" results.
* [ ] I think the code introduced in this change is easy to maintain.
* [ ] I understand all the code in this PR.

**Comment by njriasan** [comment]
> Am going to try out the check list here #8209 on this PR, hope that's ok @njriasan?
> 

It might be helpful to have a version of this in the PR template and have the author fill it out as a first step. I understand its still a work in progress, but knowing where to add code for `pysupported.rst` for example is great to have in the instructions and may reduce review time.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @njriasan. I've given this an initial review based on the noted check list, there's a few things to look at but they are mostly small. Thanks again!

**Comment by stuartarchibald** [comment]
> > Am going to try out the check list here #8209 on this PR, hope that's ok @njriasan?
> 
> It might be helpful to have a version of this in the PR template and have the author fill it out as a first step. I understand its still a work in progress, but knowing where to add code for `pysupported.rst` for example is great to have in the instructions and may reduce review time.

I agree. I think that after a bit of refinement through use, making a new PR template for Python changes and another for NumPy changes, both containing a checklist based on the changes in #8209, would be a good idea.

**Comment by stuartarchibald** [comment]
@njriasan I think the failing doc build is because readthedocs builds against the HEAD of a PR branch without merging `main`, so it's missing dependency changes adding to `main` since starting a new release cycle. If you merge `main` into this branch it should sort itself out.

File: numba/cpython/mathimpl.py [hunk 1]
```diff
@@ -451,3 +452,50 @@ def gcd(a, b):
...
+            "math.isclose(): rel_tol must be a floating point number or integer"
+        )
+
+    # Include float because overload can be inconsistent
+    # about omitted arguments
+    if not isinstance(abs_tol, (types.Float, types.Integer, types.Omitted, float)):
+        raise TypingError(
+            "math.isclose(): abs_tol must be a floating point number or integer"
+        )
+
```
[the remark this answers is not in the document]
Comment (njriasan) [reply numba/cpython/mathimpl.py:485]: So `np.complex128` is supported, but regular `complex` from Python should not be supported. I will update this code to support all types that inherit from types.Number, but do we have any way to distinguish between these?

[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/cpython/mathimpl.py:485]: hmmm, that's going to be a bit awkward, Numba currently does not differentiate between Python and NumPy scalars. Perhaps the best thing to do here is to accept `types.integer_domain|types.real_domain|{types.boolean}` and then raise a `TypingError` for complex domain? The use of a NumPy complex value in `math.isclose` raises a warning about casting complex to real in that it discards the imaginary part, this probably provides little value in comparison to just testing with the `.real` or `.imag` parts so not supporting it seems fine. What do you think?

**Pull request closed without merging**